server/dash/pages/nc_summary_comparison.py

```diff
@@ -58,6 +58,133 @@ def merge_dict(d1, d2):
+    
+    Return:
+        A dataframe with new column 'colnameDT' that is in datetime type.
+    """
+    df.loc[:, colname+"DT"] = pd.to_datetime(
+        df.loc[:, colname].str[:-4].str.split("T").str.join(" "))
+    return df
+
+def generate_filtered_dataframe(api_data_df, selected_nc, selected_request_types):
+    """Outputs the filtered dataframe based on the selected filters
+    This function takes the original dataframe "api_data_df", selected neighborhood
+    council "selected_nc", as well as the selected request type "selected_request_types" 
+    to output a dataframe with matching records.
+    Args:
+        api_data_df: full 311-request data directly access from the 311 data API.
+        selected_nc: A string argument automatically detected by Dash callback 
+        function when "selected_nc" element is selected in the layout.
+        selected_request_types: A list of strings automatically detected by Dash 
+        callback function when "selected_request_types" element is selected in the
+         layout, default None.
+    Returns:
+        pandas dataframe filtered by selected neighborhood council and request types
+    """
+    if (not selected_request_types or selected_request_types == ' ') and not selected_nc:
+        df = api_data_df  
+    elif selected_nc and (not selected_request_types or selected_request_types == ' '):
+        df = api_data_df[api_data_df["councilName"] == selected_nc]
+    elif selected_request_types != ' ':
+        df = df[df["typeName"].isin(selected_request_types)]
+    else:
+        df = api_data_df
+    return df
+
+
+def filter_bad_quality_data(df, data_quality_switch=True):
+    """Filters the dataframe based on pre-defined data quality filters.
+    This function takes the original dataframe "df" and filters out records 
+    with an outlier amount of request time to close based on the Freedman-Diaconis Rule.
+    Generally 10% of data is excluded.
+    Args:
+        df: 311-request data accessed from the API.
+        data_quality_switch: A boolean argument automatically detected by Dash. 
+            callback function when "data_quality_switch" toggle element is selected in layout.
+        
+    Returns:
+        df: filtered dataframe excluding outliers.
+        num_bins: the number of bins that will be used to plot histogram.
+        data_quality_output: a string being displayed on whether the data quality switch is on or off.
+    """
+    print("* Getting quality data.")
+    df = add_datetime_column(df, "createdDate")
+    df = add_datetime_column(df, "closedDate")
+    df.loc[:, "timeToClose"] = (df.loc[:, "closedDateDT"] - df.loc[:, "createdDateDT"]).dt.days
+    # Calculate the Optimal number of bins based on Freedman-Diaconis Rule.
+
+    # Replace empty rows with 0.0000001 To avoid log(0) error later.
+    df.loc[:, "timeToClose"] = df.loc[:, "timeToClose"].fillna(0.0000001)
+
+    # Replace negative values
```

Comment:
nit: period at end. Also, this is technically removing negative values of timeToClose, right?

Is this behavior still desired, given the results of #507?

Reply:
Yes. I think Liz's analysis basically found that majority of requests with service date < created date are the issues pointe dout by #507. So that's probably one of the data quality filter we can implement upstream.